File: ios/HybridRiveImageFactory.swift
import NitroModules
import RiveRuntime

final class HybridRiveImageFactory: HybridRiveImageFactorySpec {
  private func loadFromDataSource(_ source: DataSource) -> Promise<(any HybridRiveImageSpec)> {
    return Promise.async {
      let data = try await source.createLoader().load(from: source)
      guard let renderImage = RiveRenderImage(data: data) else {
        throw RuntimeError.error(withMessage: "Failed to decode image")
      }
      return HybridRiveImage(renderImage: renderImage, dataSize: data.count)
    }
  }

  func loadFromURLAsync(url: String) throws -> Promise<(any HybridRiveImageSpec)> {
    return loadFromDataSource(try .fromURL(string: url))
  }

  func loadFromResourceAsync(resource: String) throws -> Promise<(any HybridRiveImageSpec)> {
    return loadFromDataSource(.bundle(nameWithExtension: resource))
  }

  func loadFromBytesAsync(bytes: ArrayBufferHolder) throws -> Promise<(any HybridRiveImageSpec)> {
    return loadFromDataSource(.bytes(from: bytes))
  }
}

File: android/src/new/java/com/margelo/nitro/rive/HybridRiveImageFactory.kt
package com.margelo.nitro.rive

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise

@Keep
@DoNotStrip
class HybridRiveImageFactory : HybridRiveImageFactorySpec() {

  private fun loadFromDataSource(source: DataSource): Promise<HybridRiveImageSpec> {
    return Promise.async {
      val loader = source.createLoader()
      val data = loader.load(source)
      HybridRiveImage(data)
    }
  }

  override fun loadFromURLAsync(url: String): Promise<HybridRiveImageSpec> {
    return loadFromDataSource(DataSource.fromURL(url))
  }

  override fun loadFromResourceAsync(resource: String): Promise<HybridRiveImageSpec> {
    return loadFromDataSource(DataSource.resource(resource))
  }

  override fun loadFromBytesAsync(bytes: ArrayBuffer): Promise<HybridRiveImageSpec> {
    return loadFromDataSource(DataSource.Bytes.from(bytes))
  }
}
